### PR TITLE
Fabric: clean up RPC and support show/hide toolbar message

### PR DESCRIPTION
**Pull request by languy**
[Preview this branch](https://cosmos-explorer-preview.azurewebsites.net/pull/1680)

* Now use Promise to post message to Fabric
* Add debounce when requesting allResourceToken to fast
* Simplify schedule request token
* Modify initialization sequence: first get the connectionId from Fabric, then request token
* Add `scope` to `sendCachedMessage` in order to make the posted messages unique and avoid receiving messages for different artifact when switching artifacts
* Refactored rpc contracts to keep track of versions

### CODE REVIEW

File: src/hooks/useKnockoutExplorer.ts [hunk 1]
```diff
@@ -108,31 +105,19 @@ async function configureFabric(): Promise<Explorer> {
 
         switch (data.type) {
```
Comment (ato9000) [inline src/hooks/useKnockoutExplorer.ts:106]: Let's add support for version property

switch (`${data.type}_${data.version}`) {
    case "initialize_": {
        const fabricDatabaseConnectionInfo: FabricDatabaseConnectionInfo = {
            endpoint: data.message.endpoint,
            databaseId: data.message.databaseId,
            resourceTokens: data.message.resourceTokens as { [resourceId: string]: string },
            resourceTokensTimestamp: data.message.resourceTokensTimestamp,
        };
        explorer = await createExplorerFabric(fabricDatabaseConnectionInfo);
        resolve(explorer);

        explorer.refreshAllDatabases().then(() => {
            openFirstContainer(explorer, fabricDatabaseConnectionInfo.databaseId);
        });
        scheduleRefreshDatabaseResourceToken();
        break;
    }
    case "initialize_ fabric2": {
        explorer = createExplorerFabric(data.message);
        await scheduleRefreshDatabaseResourceToken(true);
        resolve(explorer);
        await explorer.refreshAllDatabases();
        openFirstContainer(explorer, userContext.fabricContext.databaseConnectionInfo.databaseId);
        break;
    }

**Review by ato9000 — CHANGES_REQUESTED** [review]

File: src/hooks/useKnockoutExplorer.ts (hunk 1 above)
Comment (languy) [reply src/hooks/useKnockoutExplorer.ts:110]: Added a new message and explictly keep track of the message versions.

File: src/hooks/useKnockoutExplorer.ts [hunk 2]
```diff
@@ -101,40 +102,35 @@ async function configureFabric(): Promise<Explorer> {
...
           case "initialize": {
-            const fabricDatabaseConnectionInfo: FabricDatabaseConnectionInfo = {
-              endpoint: data.message.endpoint,
-              databaseId: data.message.databaseId,
-              resourceTokens: data.message.resourceTokens as { [resourceId: string]: string },
-              resourceTokensTimestamp: data.message.resourceTokensTimestamp,
-            };
-            explorer = await createExplorerFabric(fabricDatabaseConnectionInfo);
+            const fabricVersion = data.version;
```
Comment (ato9000) [inline src/hooks/useKnockoutExplorer.ts:112]: This should be an array of versions so we can make Fabric support backward versions

**Review by ato9000 — DISMISSED** [review]

File: src/hooks/useKnockoutExplorer.ts (hunk 2 above)
Comment (languy) [reply src/hooks/useKnockoutExplorer.ts:112]: This is the version of the fabric client that is currently talking to DataExplorer. The array containing the supported versions is hardcoded in the respective clients.

File: src/Common/MessageHandler.ts [hunk 3]
```diff
@@ -54,6 +63,10 @@ export function sendMessage(data: any): void {
   });
 }
 
+/**
+ *
+ * @param data Overwrite the data property of the message
```
Comment (vchske) [inline src/Common/MessageHandler.ts:68]: Do we need to add this comment? Doesn't look like the method accepts a param.

**Review by vchske — DISMISSED** [review]

File: src/Common/MessageHandler.ts (hunk 3 above)
Comment (languy) [reply src/Common/MessageHandler.ts]: Looks like the comment was for the previous method. Thanks for the catch!

File: src/Contracts/MessageTypes.ts [hunk 4]
```diff
@@ -37,10 +43,9 @@ export enum MessageTypes {
   DisplayNPSSurvey,
   OpenVCoreMongoNetworkingBlade,
   OpenVCoreMongoConnectionStringsBlade,
-
-  // Data Explorer -> Fabric communication
-  GetAuthorizationToken,
-  GetAllResourceTokens,
+  GetAuthorizationToken, // Data Explorer -> Fabric
```
Comment (jawelton74) [inline src/Contracts/MessageTypes.ts:46]: Laurent - can we plan to split message types into Portal message types and Fabric message types in the near future ?

**Review by jawelton74 — APPROVED** [review]

File: src/Contracts/MessageTypes.ts (hunk 4 above)
Comment (languy) [reply src/Contracts/MessageTypes.ts:46]: The constraint is that we are leveraging the existing messaging infrastructure and its an enum (ie Fabric receives a number), so if we have a completely separate enum file for Fabric, we could get collisions.
Let's take this offline.

**Review by sunghyunkang1111 — APPROVED** [review]

**Pull request merged**